api-report/counter.api.md

```diff
@@ -0,0 +1,49 @@
+## API Report File for "@fluidframework/counter"
+
+> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).
+
+```ts
+
+import { IChannelFactory } from '@fluidframework/datastore-definitions';
+import { IChannelStorageService } from '@fluidframework/datastore-definitions';
+import { IFluidDataStoreRuntime } from '@fluidframework/datastore-definitions';
+import { IFluidSerializer } from '@fluidframework/core-interfaces';
+import { ISequencedDocumentMessage } from '@fluidframework/protocol-definitions';
+import { ISharedObject } from '@fluidframework/shared-object-base';
+import { ISharedObjectEvents } from '@fluidframework/shared-object-base';
+import { ITree } from '@fluidframework/protocol-definitions';
+import { SharedObject } from '@fluidframework/shared-object-base';
+
+// @public
+export interface ISharedCounter extends ISharedObject<ISharedCounterEvents> {
+    increment(incrementAmount: number): void;
+    value: number;
+}
+
+// @public (undocumented)
+export interface ISharedCounterEvents extends ISharedObjectEvents {
+    // (undocumented)
+    (event: "incremented", listener: (incrementAmount: number, newValue: number) => void): any;
+}
+
+// @public
+export class SharedCounter extends SharedObject<ISharedCounterEvents> implements ISharedCounter {
```

Comment:
I do not see much value in having SharedCounter class in here. It should be internal implementation detail., It should be free to change over time as long as public API (ISharedCounter) is not changed.
Other way to say it - this class should not be exported from package. Only interfaces, and factory.
Basically either we push developers to use interfaces, or we should ditch interfaces and export class only - there is not much value in having both.

Reply:
This makes sense to me. @ChumpChief your thoughts? If we remove the class export is it possible to rename the exported interface to `SharedCounter`?

Reply:
Couple thoughts:

I agree in principle that a developer could be successful with just the factory and `ISharedCounter`, but this is not the pattern we currently use throughout the repo (our DDS classes provide static `create()` and `getFactory()` that are used pretty extensively).  If we want to change this we should do it uniformly across all DDSs to avoid unique surprises on a per-DDS basis.

That said, I think most developers are not very happy with our pervasive use of factories.  Even though it's nice to reduce exports, I'm not confident whether pushing these scenarios to use the factory is a move in the right or wrong direction.  I'm kind of "change is bad unless it's great" here at the moment.

In any case, I would also still prefer exporting the interface as `ISharedCounter`, since I personally think the 'I' prefix is helpful.

Reply:
I filed #6399 to take up the interface vs. class discussion for DDSes.